### PR TITLE
Fix some -Wfloat-conversion warnings (Pt. 4)

### PR DESCRIPTION
**Pull request by Holzhaus**

### CODE REVIEW

File: src/effects/builtin/balanceeffect.cpp [hunk 1]
```diff
@@ -116,8 +116,8 @@ void BalanceEffect::processChannel(const ChannelHandle& handle,
     CSAMPLE_GAIN balance = 0;
     CSAMPLE_GAIN midSide = 0;
     if (enableState != EffectEnableState::Disabling) {
-        balance = m_pBalanceParameter->value();
-        midSide = m_pMidSideParameter->value();
+        balance = static_cast<CSAMPLE_GAIN>(m_pBalanceParameter->value());
+        midSide = static_cast<CSAMPLE_GAIN>(m_pMidSideParameter->value());
```
Comment (Holzhaus) [inline src/effects/builtin/balanceeffect.cpp:120]: decltype

File: src/engine/filters/enginefilterpansingle.h [hunk 2]
```diff
@@ -53,8 +53,10 @@ class EngineFilterPanSingle {
         m_delayFrame = (m_delayFrame + 1) % SIZE;
 
         // prepare coefficients for linear interpolation using a linear stretching
-        CSAMPLE_GAIN timeBetweenFullSamplesLeft = fmod(delayLeftSourceFrame, 1);
-        CSAMPLE_GAIN timeBetweenFullSamplesRight = fmod(delayRightSourceFrame, 1);
+        CSAMPLE_GAIN timeBetweenFullSamplesLeft =
+                static_cast<CSAMPLE_GAIN>(fmod(delayLeftSourceFrame, 1));
+        CSAMPLE_GAIN timeBetweenFullSamplesRight =
```
Comment (Holzhaus) [inline src/engine/filters/enginefilterpansingle.h:58]: auto

File: src/library/dlgcoverartfullsize.cpp [hunk 3]
```diff
@@ -261,8 +261,12 @@ void DlgCoverArtFullSize::wheelEvent(QWheelEvent* event) {
...
 #endif
 
-    int newPointX = (double) oldPointUnderCursor.x() / oldWidth * newSize.width();
-    int newPointY = (double) oldPointUnderCursor.y() / oldHeight * newSize.height();
+    int newPointX =
+            static_cast<int>(static_cast<double>(oldPointUnderCursor.x()) /
+                    oldWidth * newSize.width());
+    int newPointY =
+            static_cast<int>(static_cast<double>(oldPointUnderCursor.y()) /
+                    oldHeight * newSize.height());
```
Comment (Holzhaus) [inline src/library/dlgcoverartfullsize.cpp:269]: auto

File: src/effects/builtin/flangereffect.cpp [hunk 4]
```diff
@@ -222,7 +224,7 @@ void FlangerEffect::processChannel(const ChannelHandle& handle,
         CSAMPLE prevRight = delayRight[framePrev];
         CSAMPLE nextRight = delayRight[frameNext];
 
-        CSAMPLE frac = delayFrames - floorf(delayFrames);
+        auto frac = static_cast<CSAMPLE>(delayFrames - floorf(static_cast<CSAMPLE>(delayFrames)));
```
Comment (daschuer) [inline src/effects/builtin/flangereffect.cpp:227]: delayFrames can become a float, it is a position, not a CSAMPLE. 

frac is CSAMPLE_GAIN. 
Here we have a case where auto is misleading, basausse only the destination is CSAMPLE_GAIN 

The whole part can be optimized by using modf() including the floor() and ceil() calls above.

File: src/engine/filters/enginefiltermoogladder4.h [hunk 5]
```diff
@@ -135,12 +136,10 @@ class EngineFilterMoogLadderBase : public EngineObjectConstIn {
...
-                m_postGain = m_postGainNew * cross_mix
-                        + startPostGain * (1.0 - cross_mix);
-                m_kacr = m_kacrNew * cross_mix
-                        + startKacr * (1.0 - cross_mix);
-                m_k2vg = m_k2vgNew * cross_mix
-                        + startK2vg * (1.0 - cross_mix);
+                m_postGain = static_cast<float>(m_postGainNew * cross_mix +
+                        startPostGain * (1.0 - cross_mix));
+                m_kacr = static_cast<float>(m_kacrNew * cross_mix + startKacr * (1.0 - cross_mix));
+                m_k2vg = static_cast<float>(m_k2vgNew * cross_mix + startK2vg * (1.0 - cross_mix));
```
Comment (daschuer) [inline src/engine/filters/enginefiltermoogladder4.h:142]: I am unsure here. I can remember that this code is sensitive for reduced precision so wee need to test it sorrowfully.

File: src/track/replaygain.cpp [hunk 6]
```diff
@@ -120,7 +120,7 @@ CSAMPLE ReplayGain::peakFromString(QString strPeak, bool* pValid) {
         return kPeakUndefined;
     }
     isValid = false;
-    const CSAMPLE peak = normalizedPeak.toDouble(&isValid);
+    const auto peak = static_cast<CSAMPLE>(normalizedPeak.toDouble(&isValid));
```
Comment (daschuer) [inline src/track/replaygain.cpp:123]: toFloat()

File: src/waveform/renderers/waveformrenderersignalbase.cpp [hunk 7]
```diff
@@ -154,28 +154,31 @@ void WaveformRendererSignalBase::getGains(float* pAllGain, float* pLowGain,
                                           float* pMidGain, float* pHighGain) {
     WaveformWidgetFactory* factory = WaveformWidgetFactory::instance();
     if (pAllGain != NULL) {
-        float allGain = m_waveformRenderer->getGain();
+        double allGain = m_waveformRenderer->getGain();
         allGain *= factory->getVisualGain(::WaveformWidgetFactory::All);
-        *pAllGain = allGain;
+        *pAllGain = static_cast<CSAMPLE_GAIN>(allGain);
```
Comment (daschuer) [inline src/waveform/renderers/waveformrenderersignalbase.cpp:159]: this can become float math.

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you.

File: src/effects/builtin/echoeffect.cpp [hunk 8]
```diff
@@ -197,8 +198,8 @@ void EchoEffect::processChannel(const ChannelHandle& handle, EchoGroupState* pGr
         CSAMPLE bufferedSampleLeft = gs.delay_buf[read_position];
         CSAMPLE bufferedSampleRight = gs.delay_buf[read_position + 1];
         if (read_position != prev_read_position) {
-            double frac = static_cast<double>(i)
-                / bufferParameters.samplesPerBuffer();
+            const auto frac = static_cast<CSAMPLE>(static_cast<double>(i) /
+                    bufferParameters.samplesPerBuffer());
```
[the remark this answers is not in the document]
Comment (Holzhaus) [reply src/effects/builtin/echoeffect.cpp:202]: The whole right-hand side is wrapped in a `static_cast`. Hence, I think `auto` makes sense.

File: src/engine/filters/enginefiltermoogladder4.h (hunk 5 above)
Comment (Holzhaus) [reply src/engine/filters/enginefiltermoogladder4.h:142]: This doesn't change anything. The members are already floats, so the typecast only makes this visible.

File: src/effects/builtin/echoeffect.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (daschuer) [reply src/effects/builtin/echoeffect.cpp:202]: Yes, but not if you cast only i

File: src/engine/filters/enginefiltermoogladder4.h (hunk 5 above)
Comment (daschuer) [reply src/engine/filters/enginefiltermoogladder4.h:142]: This line is OK. I am concerned about everything. 
If the filter still sounds as before everything is OK.

File: src/effects/builtin/flangereffect.cpp (hunk 4 above)
Comment (Holzhaus) [reply src/effects/builtin/flangereffect.cpp:226]: Please use the suggestion feature or open a new PR. I don't want to refactor code I don't understand.

File: src/effects/builtin/echoeffect.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (Holzhaus) [reply src/effects/builtin/echoeffect.cpp:202]: I don't understand... please use the suggestion feature.

File: src/effects/builtin/flangereffect.cpp (hunk 4 above)
Comment (daschuer) [reply src/effects/builtin/flangereffect.cpp:226]: OK, than just remove auto and make delayFrames float.

File: src/effects/builtin/echoeffect.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (daschuer) [reply src/effects/builtin/echoeffect.cpp:202]:
```suggestion
                    const CSAMPLE_GAIN frac = static_cast<CSAMPLE_GAIN>(i) /
                    bufferParameters.samplesPerBuffer();
```

File: src/engine/filters/enginefiltermoogladder4.h (hunk 5 above)
Comment (Holzhaus) [reply src/engine/filters/enginefiltermoogladder4.h:142]: Sounds okay to me, but I'm on my laptop and the speaker don't sound good anyway :wink: Please double-check.

Comment (daschuer) [reply src/engine/filters/enginefiltermoogladder4.h:142]: I can confirm this.

**Review by daschuer — APPROVED** [review]

**Comment by daschuer** [comment]
Thank you for, this kind of no fun work.

**Comment by Holzhaus** [comment]
> Thank you for, this kind of no fun work.

True. Thanks for reviewing and merging!

**Pull request merged**